### ISSUE TITLE
How to run metadynamics in opennmm using radius of gyration（Rg） as CV 

### ISSUE THREAD

**Issue by pearlDingzhen** [openmm/openmm#4095]
I want to use metadynamic to accelerate the radius of gyration for sampling a series of IDPs. What I want to ask is how this can be implemented in openMM?
1. can openMM get Rg directly? If not, can it read the Rg output from MDAnalysis or MDtraj as input to the potential energy term?
2. Metadynamic needs a BiasVariable object, and generating BiasVariable needs a force object, how should I define a CustomForce related to the radius of gyration?

**Comment by craabreu** [comment]
@pearlDingzhen, take a look at [CVPack](https://github.com/RedesignScience/cvpack).

**Comment by pearlDingzhen** [comment]
> @pearlDingzhen, take a look at [CVPack](https://github.com/RedesignScience/cvpack).

thank you. 
I tried and successfully built a Metadyhnamics object with Rg as a BiasVariable.
But now I failed to generate a simulation object. The pyscript gets stuck while generating the simulation object, no error is reported, but it doesn't proceed either.
I am puzzled and looking for some advice. Here is my code, it just print(2) and nothing else throwed out. The simulation can be generated without meta for this system
``
	system = top.create_system(nonbonded_cutoff=1.2 * nanometer)	
	peptide =  [ a.index for a in top.topology.atoms() if (a.residue.name != 'HOH' and a.residue.name != 'ION') ]
	radius_of_gyration = cvpack.RadiusOfGyration(peptide, pbc=True, weighByMass=False)	
	print (radius_of_gyration)
	grid_min = 0.0
	grid_max = 6.0
	biaswidth = 0.004
	Rg_cv = BiasVariable(radius_of_gyration, grid_min, grid_max, biaswidth, False, gridWidth=100)
	meta = Metadynamics(system, [Rg_cv], ref_t * kelvin, 4, 3*kilocalories_per_mole, 400, biasDir='./Bias/', saveFrequency= 4000)
	print('meta')
	integrator = LangevinIntegrator(ref_t * kelvin,
									10.0 / picosecond,
									time_step * picosecond)
	integrator.setRandomNumberSeed(0)

	print(2)

	simulation = Simulation(top.topology, system, integrator,
							platform, properties)
	print(3)
	simulation.context.setPositions(conf.getPositions())

**Comment by peastman** [comment]
CVPack is implementing the radius of gyration with a CustomCentroidBondForce like this:

```python
num_groups = num_atoms + 1
sum_dist_sq = "+".join([f"distance(g{i+1}, g{num_atoms + 1})^2" for i in range(num_atoms)])
super().__init__(num_groups, f"sqrt(({sum_dist_sq})/{num_atoms})")
for atom in group:
    self.addGroup([atom], [1])
self.addGroup(group, None if weighByMass else [1] * num_atoms)
self.addBond(list(range(num_groups)), [])
```

It creates a single bond that depends on every atom all at once, with a huge energy expression containing a separate term for every atom.  That's ok for just a few atoms, but for large numbers of atoms it will be insanely expensive.

Here is a more efficient way to do it.  Add one group for each atom, plus a final group that computes the center of mass.  Add a bond for each atom that computes its distance to the center of mass, something like this.

```python
super().__init__(2, f"(distance(g1, g2)^2)/{num_atoms}")
for atom in group:
    self.addGroup([atom], [1])
self.addGroup(group, None if weighByMass else [1] * num_atoms)
for atom in group:
    self.addBond([atom, num_atoms])
```

That will give you the square of the radius of gyration, which may be good enough for some purposes.  If you need to take the square root, you can do it by wrapping it in a CustomCVForce.

**Comment by craabreu** [comment]
It is true that biasing `rg^2` instead of `rg` can be good enough. But when one actually needs to transform a CV before applying the Metadynamics bias, wrapping it in a `CustomCVForce` would result in nested `CustomCVForce` objects. Allowing the user to define a transformation when instantiating a `BiasVariable` could avoid this situation. For example, it could be passed as a lepton string defining a simple function of `x` (defaulting to `x` itself, of course).

**Comment by peastman** [comment]
Good idea.  That could be implemented by just changing the expression the Metadynamics class creates for its own CustomCVForce.

There's nothing wrong with nesting CustomCVForces, except that it adds a bit more overhead.

**Comment by craabreu** [comment]
For future reference, a new [RadiusOfGyrationSq](https://redesignscience.github.io/cvpack/api/RadiusOfGyrationSq.html) collective variable has been added to CVPack as the preferential choice for large groups of atoms. Thanks for the suggestion, @peastman.

**Comment by peastman** [comment]
Looks great.

**Comment by dsclassen** [comment]
... and what might be considered a "large group of atoms"? I've been attempting to implement `RadiusOfGyrationSq` for various protein systems with several 1000 atoms and the script just hangs when trying to instantiate the integrator. If I just hardcode the first 50-100 atom indices then things are working. I assume that the calculations are some sort of O(N²) complexity?

**Comment by peastman** [comment]
It's probably just trying to optimize your huge expression.  Even if you wait for it to get through doing that, it will run very slowly because you're forcing it to do the entire calculation in a single thread.

**Comment by peastman** [comment]
OpenMM 8.4 adds RGForce, which is an efficient way of computing it.
